### PR TITLE
feat(ticket#49):  Leave Request Action Implementation 

### PR DESCRIPTION
**Pull request by nadirhussain786**
## Description 

This PR contain server action for leave request.

### Request Body

```
type Attendance = {
id: number;
memberId: number;
shift: string
startTime: string;
endTime: string;
status: ['pending' , 'rejected', 'approved']; // in case of creation our status will always 'pending'
}
```

### Resources

image

### Test

`action.test.tsx` file contain server action test. In this test  `onCreateLeaveRequest` function submit with mock payload.

### CODE REVIEW

File: src/app/leaves/actions/actions.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+"use server";
+import { z } from "zod";
+import { getSupabaseClient } from "../../../../utils/supabase/supabaseClient";
+import { leaveRequestSchema } from "../add/components/LeaveRequestFormBuilder";
+import { redirect } from "next/navigation";
+import { LeaveRequestStatus } from "@/constant/constant";
+
+export const onCreateLeaveRequest = async (
+  values: z.infer<typeof leaveRequestSchema>
```
Comment (Mubashar1009) [inline src/app/leaves/actions/actions.ts:9]: can we use type here

File: src/app/leaves/actions/action.test.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  it("calls onCreateLeaveRequest with mock form data on submit", async () => {
+    const submit = vi.fn().mockImplementation(onCreateLeaveRequest);
+    await waitFor(() => {
+      submit({
+        ...mockPayload,
+      });
+    });
+    expect(submit).toHaveBeenCalledTimes(1);
+  });
+});
```
Comment (Mubashar1009) [inline src/app/leaves/actions/action.test.ts:24]: can we use this in the LeaveRequestFormBuilder.test file so it would be then integeration testing

**Review by ghulamhafeez — APPROVED** [review]

File: src/app/leaves/actions/action.test.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import { waitFor } from "@testing-library/dom";
```
Comment (Monfernape) [inline src/app/leaves/actions/action.test.ts:1]: File name is wrong. It should depict the action name.

File: src/app/leaves/actions/actions.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+"use server";
```
Comment (Monfernape) [inline src/app/leaves/actions/actions.ts:1]: Let's update file name. How about we rename out action to `createLeaveRequest`?

File: src/app/leaves/actions/actions.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import { z } from "zod";
+import { getSupabaseClient } from "../../../../utils/supabase/supabaseClient";
+import { leaveRequestSchema } from "../add/components/LeaveRequestFormBuilder";
+import { redirect } from "next/navigation";
+import { LeaveRequestStatus } from "@/constant/constant";
+
+export const onCreateLeaveRequest = async (
```
Comment (Monfernape) [inline src/app/leaves/actions/actions.ts:8]: Change this name too. We only add `on` prefix with functions which are event handlers or callbacks.

File: src/app/leaves/actions/actions.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  values: z.infer<typeof leaveRequestSchema>
+) => {
+
+  const supabase = await getSupabaseClient();
+  const { error } = await supabase.from("leaves").insert({
```
Comment (Monfernape) [inline src/app/leaves/actions/actions.ts:13]: Can we use enums for tables here?

File: src/app/leaves/actions/actions.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    endDate: values.endDate.toISOString(),
+    reasonForLeave: values.reason,
+    status: LeaveRequestStatus.Pending,
+  });
+
+  if (error) {
+    return error;
+  }
+
+  redirect("/leaves");
```
Comment (Monfernape) [inline src/app/leaves/actions/actions.ts:26]: Route enum here as well.

File: src/app/leaves/add/components/LeaveRequestFormBuilder.tsx [hunk 8]
```diff
@@ -99,7 +112,7 @@ export const LeaveRequestFormBuilder = () => {
       </h1>
       <Form {...form}>
         <form
-          data-testid="form"
+          data-testid="validForm"
```
Comment (Monfernape) [inline src/app/leaves/add/components/LeaveRequestFormBuilder.tsx:115]: This is a weird id. Should represent form name.

File: src/constant/constant.ts [hunk 9]
```diff
@@ -23,3 +23,9 @@ export enum LeaveTypes {
   Vacation = "vacation",
   Personal = "personal",
 }
+
+export enum LeaveRequestStatus {
+  Pending = "pending",
+  Accept = "accept",
```
Comment (Monfernape) [inline src/constant/constant.ts:29]: Correct terminology would be accepted and rejected.

File: src/app/leaves/actions/actions.ts (hunk 1 above)
Comment (nadirhussain786) [reply src/app/leaves/actions/actions.ts:9]: I have implemented this improvement.

File: src/app/leaves/actions/action.test.ts (hunk 3 above)
Comment (nadirhussain786) [reply src/app/leaves/actions/action.test.ts:1]: Fixed this one

File: src/app/leaves/actions/actions.ts (hunk 4 above)
Comment (nadirhussain786) [reply src/app/leaves/actions/actions.ts:1]: Fixed

File: src/app/leaves/actions/actions.ts (hunk 6 above)
Comment (nadirhussain786) [reply src/app/leaves/actions/actions.ts:13]: Yes, I have fixed this by creating an enum.

File: src/app/leaves/actions/actions.ts (hunk 7 above)
Comment (nadirhussain786) [reply src/app/leaves/actions/actions.ts:26]: Fixed by using `Routes` enum

File: src/app/leaves/add/components/LeaveRequestFormBuilder.tsx (hunk 8 above)
Comment (nadirhussain786) [reply src/app/leaves/add/components/LeaveRequestFormBuilder.tsx:115]: Fixed

File: src/constant/constant.ts (hunk 9 above)
Comment (nadirhussain786) [reply src/constant/constant.ts:29]: Fixed these terminology

**Pull request merged**